connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/WorkerCoordinator.java

```diff
@@ -145,154 +166,79 @@ public void poll(long timeout) {
     }
 
     @Override
-    public JoinGroupRequestData.JoinGroupRequestProtocolSet metadata() {
+    public JoinGroupRequestProtocolSet metadata() {
         configSnapshot = configStorage.snapshot();
-        ConnectProtocol.WorkerState workerState = new ConnectProtocol.WorkerState(restUrl, configSnapshot.offset());
-        ByteBuffer metadata = ConnectProtocol.serializeMetadata(workerState);
-        return new JoinGroupRequestData.JoinGroupRequestProtocolSet(
-                Collections.singleton(new JoinGroupRequestData.JoinGroupRequestProtocol()
-                        .setName(DEFAULT_SUBPROTOCOL)
-                        .setMetadata(metadata.array()))
-                        .iterator()
-        );
+        ExtendedWorkerState workerState = new ExtendedWorkerState(restUrl, configSnapshot.offset(), assignmentSnapshot);
+        ByteBuffer metadata;
+        switch (protocolCompatibility) {
+            case EAGER:
+                metadata = ConnectProtocol.serializeMetadata(workerState);
+                return new JoinGroupRequestProtocolSet(Collections.singleton(
+                        new JoinGroupRequestProtocol()
+                                .setName(protocolCompatibility.protocol())
+                                .setMetadata(metadata.array()))
+                        .iterator());
+            case COMPATIBLE:
+                return new JoinGroupRequestProtocolSet(Arrays.asList(
+                        new JoinGroupRequestProtocol()
+                                .setName(protocolCompatibility.protocol())
+                                .setMetadata(IncrementalCooperativeConnectProtocol.serializeMetadata(workerState).array()),
+                        new JoinGroupRequestProtocol()
+                                .setName(EAGER.protocol())
+                                .setMetadata(ConnectProtocol.serializeMetadata(workerState).array()))
+                        .iterator());
+            default:
+                throw new IllegalStateException("Unknown Connect protocol compatibility mode " + protocolCompatibility);
+        }
     }
 
     @Override
     protected void onJoinComplete(int generation, String memberId, String protocol, ByteBuffer memberAssignment) {
-        assignmentSnapshot = ConnectProtocol.deserializeAssignment(memberAssignment);
+        ConnectAssignment newAssignment = IncrementalCooperativeConnectProtocol.deserializeAssignment(memberAssignment);
```

Comment:
I'm still working through updated code and all the classes, but this feels a bit messy. Admittedly we just dumped the `ConnectProtocol` here initially when we only had one type. But now it seems confusing when I use the new protocol directly vs having to switch on compatibility, etc. Not sure I have a concrete suggestion yet, but I worry about the more general classes here ending up with both explicit (compat mode) and implicit (this case where we always use `IncrementalCooperativeConnectProtocol`) decisions about how to handle logic of different versions.

I'll think more about how we can refactor...

Reply:
Again here the difficulty is part due to protocol classes being effectively static. Since it's not critical for the initial version of the new rebalancing, let's revisit in a follow-up.